File: packages/expo-speech/android/src/main/java/expo/modules/speech/LanguageUtils.java
package expo.modules.speech;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

// Lazy load the ISO codes into a Map then transform the codes to match other localization patterns in Expo
public class LanguageUtils {

  private static Map<String, Locale> countryISOCodes;

  private static Map<String, Locale> languageISOCodes;

  private static String transformCountryISO(String code) {
    if (countryISOCodes == null) {
      String[] countries = Locale.getISOCountries();
      countryISOCodes = new HashMap<>(countries.length);
      for (String country : countries) {
        Locale locale = new Locale("", country);
        countryISOCodes.put(locale.getISO3Country().toUpperCase(), locale);
      }
    }
    return countryISOCodes.get(code).getCountry();
  }

  private static String transformLanguageISO(String code) {
    if (languageISOCodes == null) {
      String[] languages = Locale.getISOLanguages();
      languageISOCodes = new HashMap<>(languages.length);
      for (String language : languages) {
        Locale locale = new Locale(language);
        languageISOCodes.put(locale.getISO3Language(), locale);
      }
    }
    return languageISOCodes.get(code).getLanguage();
  }

  static String getISOCode(Locale locale) {
    String language = transformLanguageISO(locale.getISO3Language());
    String country = locale.getISO3Country();
    if (!country.equals("")) {
      String countryCode = transformCountryISO(country);
      language += "-" + countryCode;
    }
    return language;
  }
}

File: packages/expo-speech/android/src/main/java/expo/modules/speech/LanguageUtils.kt
package expo.modules.speech

import java.util.*

// Lazy load the ISO codes into a Map then transform the codes to match other localization patterns in Expo
object LanguageUtils {
  private val countryISOCodes: Map<String, Locale> by lazy {
    Locale.getISOCountries().associate { country ->
      val locale = Locale("", country)
      locale.isO3Country.uppercase(locale) to locale
    }
  }
  private val languageISOCodes: Map<String, Locale> by lazy {
    Locale.getISOLanguages().associate { language ->
      val locale = Locale(language)
      locale.isO3Language to locale
    }
  }

  fun getISOCode(locale: Locale): String {
    val language =
      try {
        languageISOCodes[locale.isO3Language]?.language
      } catch (_: MissingResourceException) {
        null
      } ?: locale.language

    val country =
      try {
        locale.isO3Country.takeIf { it != "" }?.let { countryISOCodes[it]?.country }
      } catch (_: MissingResourceException) {
        null
      } ?: locale.country

    return if (country.isNotEmpty()) "$language-$country" else language
  }
}
